java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -524,34 +525,50 @@ public void visitLambdaExpression(LambdaExpressionTree tree) {
   }
 
   private void refineLambdaType(LambdaExpressionTreeImpl lambdaExpressionTree, JavaType lambdaType) {
-    JavaType samReturnType = (JavaType) getSamMethod(lambdaType).returnType().type();
+    JavaSymbol.MethodJavaSymbol samMethod = getSamMethod(lambdaType);
+    if(samMethod == null) {
+      return;
+    }
+    JavaType samReturnType = (JavaType) samMethod.returnType().type();
     JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, lambdaType);
     if (lambdaExpressionTree.body().is(Tree.Kind.BLOCK) || capturedReturnType.is("void")) {
       // TODO compute return type for lambda with block and return statements
       return;
     }
     JavaType refinedReturnType = (JavaType) ((AbstractTypedTree) lambdaExpressionTree.body()).symbolType();
-    if (refinedReturnType != capturedReturnType && lambdaType.isParameterized()) {
+    refineType(lambdaExpressionTree, lambdaType, capturedReturnType, refinedReturnType);
+  }
+
+  private void refineType(AbstractTypedTree expression, JavaType expressionType, JavaType capturedReturnType, JavaType refinedReturnType) {
+    if (refinedReturnType != capturedReturnType && expressionType.isParameterized()) {
       // found a lambda return type different from the one infered : update infered type
-      TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) lambdaType).typeSubstitution;
+      TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) expressionType).typeSubstitution;
       for (Map.Entry<TypeVariableJavaType, JavaType> entry : typeSubstitution.substitutionEntries()) {
         if (entry.getValue() == capturedReturnType) {
           TypeSubstitution refinedSubstitution = new TypeSubstitution(typeSubstitution).add(entry.getKey(), refinedReturnType);
-          JavaType refinedLambdaType = parametrizedTypeCache.getParametrizedTypeType(lambdaType.symbol, refinedSubstitution);
-          lambdaExpressionTree.setType(refinedLambdaType);
+          JavaType refinedType = parametrizedTypeCache.getParametrizedTypeType(expressionType.symbol, refinedSubstitution);
+          expression.setType(refinedType);
           break;
         }
       }
     }
   }
 
+  @CheckForNull
   private static JavaSymbol.MethodJavaSymbol getSamMethod(JavaType lambdaType) {
     for (Symbol member : lambdaType.symbol().memberSymbols()) {
       if (member.isMethodSymbol() && member.isAbstract()) {
         return (JavaSymbol.MethodJavaSymbol) member;
       }
     }
-    throw new IllegalStateException("Method not found in SAM Interface : "+lambdaType.fullyQualifiedName());
+    for (ClassJavaType type : lambdaType.symbol.superTypes()) {
+      for (Symbol member : type.symbol().memberSymbols()) {
+        if (member.isMethodSymbol() && member.isAbstract()) {
+          return (JavaSymbol.MethodJavaSymbol) member;
+        }
+      }
+    }
+    return null;
```

Comment:
It can be expected for method reference : foo(a::toString) the to String is not a sam method.